File: netDxf/IO/DxfWriter.SectionSettings.cs
using System;
using netDxf.Header;
using netDxf.Objects;

namespace netDxf.IO
{
    internal sealed partial class DxfWriter
    {
        private bool WriteSectionSettingsPayload(DxfDatabaseObject item)
        {
            if (!(item is DxfSectionSettings settings)) return false;
            if (this.doc.DrawingVariables.AcadVer < DxfVersion.AutoCad2007) throw new NotSupportedException("Typed SECTIONSETTINGS requires DXF 2007 or later.");
            settings.ValidateValues();
            this.chunk.Write(100, "AcDbSectionSettings"); this.chunk.Write(90, settings.SectionType); this.chunk.Write(91, settings.TypeSettings.Count);
            foreach (DxfSectionTypeSettings type in settings.TypeSettings)
            {
                this.chunk.Write(1, "SectionTypeSettings"); this.chunk.Write(90, type.SectionType); this.chunk.Write(91, type.GenerationOptions);
                this.chunk.Write(92, type.SourceObjects.Count);
                foreach (DxfObject source in type.SourceObjects) this.chunk.Write(330, source?.Handle ?? "0");
                this.chunk.Write(331, type.DestinationBlock?.Handle ?? "0"); this.chunk.Write(1, this.EncodeDatabaseString(type.DestinationFileName));
                this.chunk.Write(93, type.GeometrySettings.Count);
                if (!type.RepeatGeometryMarkers) this.chunk.Write(2, "SectionGeometrySettings");
                foreach (DxfSectionGeometrySettings geometry in type.GeometrySettings)
                {
                    if (type.RepeatGeometryMarkers) this.chunk.Write(2, "SectionGeometrySettings");
                    this.chunk.Write(90, geometry.SectionType); this.chunk.Write(91, geometry.GeometryValue); this.chunk.Write(92, geometry.Flags);
                    this.chunk.Write(geometry.ColorCode, geometry.ColorIndex);
                    this.chunk.Write(8, this.EncodeDatabaseString(geometry.LayerName)); this.chunk.Write(6, this.EncodeDatabaseString(geometry.LinetypeName));
                    this.chunk.Write(40, geometry.LinetypeScale); this.chunk.Write(1, this.EncodeDatabaseString(geometry.PlotStyleName)); this.chunk.Write(370, geometry.Lineweight);
                    this.chunk.Write(70, geometry.FaceTransparency); this.chunk.Write(71, geometry.EdgeTransparency); this.chunk.Write(72, geometry.HatchPatternType);
                    this.chunk.Write(2, this.EncodeDatabaseString(geometry.HatchPatternName)); this.chunk.Write(41, geometry.HatchAngle);
                    this.chunk.Write(42, geometry.HatchScale); this.chunk.Write(43, geometry.HatchSpacing); this.chunk.Write(3, "SectionGeometrySettingsEnd");
                }
                this.chunk.Write(3, "SectionTypeSettingsEnd");
            }
            return true;
        }
    }
}

File: javascript/netDxf/IO/DxfWriter.SectionSettings.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { DxfSectionSettings } from '../Objects/DxfSectionSettings.js';
import { EncodeDxfDatabaseText } from '../../runtime/DxfStringEncoding.js';
import { NotSupportedException } from '../../runtime/Errors.js';
export function WriteSectionSettingsPayload(chunk,version,item) {
  if(!(item instanceof DxfSectionSettings))return false;
  if(version<15)throw new NotSupportedException('Typed SECTIONSETTINGS requires DXF 2007 or later.');
  item.ValidateValues();const text=value=>EncodeDxfDatabaseText(value,version);
  chunk.Write(100,'AcDbSectionSettings');chunk.Write(90,item.SectionType);chunk.Write(91,item.TypeSettings.Count);
  for(const type of item.TypeSettings) {
    chunk.Write(1,'SectionTypeSettings');chunk.Write(90,type.SectionType);chunk.Write(91,type.GenerationOptions);
    chunk.Write(92,type.SourceObjects.Count);for(const source of type.SourceObjects)chunk.Write(330,source?.Handle??'0');
    chunk.Write(331,type.DestinationBlock?.Handle??'0');chunk.Write(1,text(type.DestinationFileName));chunk.Write(93,type.GeometrySettings.Count);
    if(!type.RepeatGeometryMarkers)chunk.Write(2,'SectionGeometrySettings');
    for(const value of type.GeometrySettings) {
      if(type.RepeatGeometryMarkers)chunk.Write(2,'SectionGeometrySettings');
      chunk.Write(90,value.SectionType);chunk.Write(91,value.GeometryValue);chunk.Write(92,value.Flags);
      chunk.Write(value.ColorCode,value.ColorIndex);chunk.Write(8,text(value.LayerName));chunk.Write(6,text(value.LinetypeName));
      chunk.Write(40,value.LinetypeScale);chunk.Write(1,text(value.PlotStyleName));chunk.Write(370,value.Lineweight);
      chunk.Write(70,value.FaceTransparency);chunk.Write(71,value.EdgeTransparency);chunk.Write(72,value.HatchPatternType);
      chunk.Write(2,text(value.HatchPatternName));chunk.Write(41,value.HatchAngle);chunk.Write(42,value.HatchScale);chunk.Write(43,value.HatchSpacing);
      chunk.Write(3,'SectionGeometrySettingsEnd');
    }
    chunk.Write(3,'SectionTypeSettingsEnd');
  }
  return true;
}
